### ISSUE TITLE
Replacing "deriving" with "ppx_deriving"

### ISSUE THREAD

**Issue by dhil** [links-lang/links#30]
Links does not compile with the latest OCaml compiler (4.03), because the package "deriving" is not compatible with 4.03 (actually I think it is one of the dependencies of "deriving" that's causing the issue). 

The package "deriving" appears to be obsolete, and for future projects with Links it may very well be worthwhile replacing "deriving" with "ppx_deriving". My own project would benefit from this change as well as the OCaml Multicore backend will be released for 4.03+.

According to Sam switching from "deriving" to "ppx_deriving" is not an entirely mechanical process. Furthermore, the proposed change is likely to be thorough, and since it may benefit Links in general James suggested that we implement this change on the de facto master-branch "sessions". 

The question remains: _Which parts of Links are sensitive to this change?_

**Comment by jstolarek** [comment]
Do the build errors look like [this](https://gist.github.com/jstolarek/0b2f6daf79726413b6323d7cf1b99570)? That's my build log using 4.03.0 compiler. I wonder whether I am affected by this ticket or if it's something else.

> the package "deriving" is not compatible with 4.03

How does incompatibility with 4.03 manifest? I did `opam install deriving` and that seems to have worked fine.

**Comment by slindley** [comment]
A more recent release of deriving (0.7.1) does now work with ocaml 4.03.0. It might well be worth switching to ppx_deriving at some point, but it probably isn't urgent.

@jstolarek: the problem you're having seems to relate to deriving - that is what the build errors looked like when deriving wasn't installed, so its strange that you get that now

**Comment by jstolarek** [comment]
@slindley I just tried again and I was able to build Links with 4.03.1. I have absolutely no idea why it failed previously...

**Comment by jamescheney** [comment]
Is this issue still worth doing, or should we close it?

**Comment by slindley** [comment]
Absolutely it's worth doing as ppx plays much more nicely with the OCaml toolchain than ocamlp4 does.

**Comment by jamescheney** [comment]
OK.  See also #79.

**Comment by dhil** [comment]
I recently discovered that deriving does not support all modern OCaml syntax. In particular, I found that implicit functor syntax is not supported, i.e.
```ocaml
module M = functor (X : A) (Y : B) -> struct ... end
```
has to be written
```ocaml
module M = functor (X : A) -> functor (Y : B) -> struct ... end
```
Thus I wonder what other syntax deriving does not support. Furthermore, as the OCaml community is waging war against camlp4 I think we should start getting rid of deriving in our code base.

**Comment by jamescheney** [comment]
This is increasingly critical given OCaml 4.0.6 changes and lack of maintenance of `deriving`.  We should do this for 0.8.

**Comment by dhil** [comment]
@dbuenzli suggested that it might be worth trying `camlp4-to-ppx` to mechanically derive `deriving` as a ppx extension (https://github.com/janestreet/camlp4-to-ppx).

**Comment by frank-emrich** [comment]
I'm currently working on this. I'm halfway through porting uses of Show and Eq, because there are only minor adaptions necessary which are usually straightforward

What's more involved seems to be porting uses of Typeable, Dump, and Pickle. Dump and Pickle both provide (de)serialisation and need Typeable to do so.
It seems like the only usage of this functionality in our code is in the following functions in value.ml:
val marshal_value : t -> string
val marshal_continuation : continuation -> string
val unmarshal_continuation : env -> string -> continuation
val unmarshal_value : env -> string -> t

As far as I can tell, the (un)marshalling code is only used to send stuff to the frontend. In addition, there is already a mechanism in place that allows different (un)marshalling engines to be used. Right now, a value in the settings allows to choose between Dump and Pickle (the serialisation functionality provided by the old deriving preprocessor) and OCaml's builtin Marshall module. Right now, the default engine to be used is Dump.

There is no direct counterpart to Dump and Pickle in ppx_deriving, however, there is ppx_deriving_yojson, which allows the generation of code that (un)marshalls to json. Unfortunately, I don't know what the motivation was to make multiple marshalling engines available in Links. I guess if the builtin Marshall would have worked flawlessly, we would just be using that one? So I guess that when I replace uses of Dump and Pickle by ppx_deriving_yojson, there is a possibility to run into similar problems as the ones causing people to move away from Marshall.


Also, it doesn't seem like there is a migration path that allows us to use ppx_deriving while using the old deriving preprocessor to provide Dump and Pickle. Jbuilder allows specifying  the preprocessor to be used on a per-module basis. However, those types that are annotated with Dump and Pickle are not disjoint from those for which we want Eq and Show. This means that we can't just use ppx_deriving for Show and Eq while keeping the old deriving preprocessor for Dump and Pickle.

In conclusion, I would suggest to try replacing Dump and Pickle with ppx_deriving_yojson and see if that works. Does anyone know if the frontend relies on the marshalled values from the server to be in a specific format? Or is it sufficient if there is a pair of matching marshalling and unmarshalling functions on the server and that's it?

**Comment by frank-emrich** [comment]
Good news! I was able to finish the transformation as a whole. So far everything looks to work nicely, including using ppx_deriving_yojson  instead of Dump/Pickle.
I'll double-check in the next days that I've correctly translated all customizations and clean up a bit. I should be able to create a PR by the end of next week.
